### ISSUE TITLE
First impressions

### ISSUE THREAD

**Issue by haf** [GoogleContainerTools/skaffold#1378]
This is a review of the initial experience using you tool to manage this for dev:

 - one API service
 - one backend batch service
 - one data database: eventstore
 - one metrics database: influxdb
 - prometheus

The first experience is one of an MVP that ticks the most important box; a quick getting started doc that actually works. That said, the quick start seems to have displaced the architecture diagrams and overview docs; something DevSpace did really well.

After copying some staging yaml files from our k8s repo to the local dev repo, I note that there's no guidance on how to think about helm charts for monitoring (which we need in dev), versus helm charts versus plain yaml files. Missing docs for this. Asking on [stackoverflow](https://stackoverflow.com/questions/53672777/deploy-external-helm-chart-with-skaffold/53677969#53677969) doesn't yield many responses, and the only one I've gotten so far doesn't help; I want to declare both ephemeral things (the containers with the code I change), and more stable things like Prometheus in a single file.

Since there's no architecture diagram anywhere, I don't know how Skaffold is supposed to sit between dev and prod. It would seem it's mostly a wrapper that creates/applies/deletes all yaml files I have in a folder. It's not a full answer to a development environment.

After getting the services up... (more or less; my deployment is properly rebuilt, but the backend statefulset doesn't; it tries to pull from my local GCP repo (which I have authenticated towards), despite skaffold actually building the image before it starts the statefulset — it would seem skaffold doesn't properly update the image at runtime for stateful set, but does properly do it for deployments)... I try to request the service; another thing DevSpace does much better — what use is a dev tool if I can't request things on it? I have to search for how to set up ingresses with minikube and then search for how to add TLS to it. This should be in the docs IMO.

Then, the sync; DevSpace spawns a shell inside a container — a nice touch — skaffold just starts it all and expect you to write down an exhaustive pattern match of files to sync. In my case, I want to keep the databases' state on host disk (like docker compose does with volumes); but there are no docs about this and nothing in the annotated skaffold yaml file either.

This project is heading in the right direction; but only because what tiny feature set it already has works as it should (the creation, updating and deletion of resources). There's a whole bunch of docs missing and it's unclear what the final mission of the project is (to a newcomer at least).

In short, this is needed:

 - Docs and links for rudimentary routing of gRPC/HTTP/HTTPS/whatever ports
 - How-to on sharing host folders with pods/containers
 - An architecture/overview diagram
 - A sample that hosts a MySQL or the like
 - Docs on how to add things that aren't ephemeral (like monitoring, istio, calico etc)
 - Guidance on whether to use Helm charts of K8S yaml files and when/how either is better.
 - **The ability to run a script before building the dockerfiles**
 - **The ability to run a script on e.g. 'api is healthy', 'db is healthy' to seed dev data/permissions**
 - Focus more on the dev experience and less on having a single file for GCP and dev.

Equivalent for DevSpace: https://github.com/covexo/devspace/issues/417

**Comment by pscarey** [comment]
Not a maintainer/affiliated, but just my 2 cents. We picked up `skaffold` this last week & have had a very different experience.

> I want to declare both ephemeral things (the containers with the code I change), and more stable things like Prometheus in a single file.

We manage this by using a helm chart with a `requirements.yaml`, and with local copies of subcharts. We can toggle the requirements with the `condition` option, allowing us to enable certain things only in prod/dev (using `skaffold`'s `profile` settings). Helm doesn't play nice with a mix of local and remote subcharts at the minute (I believe there's an issue filed over this in the Helm repo). Skaffold only needs to be told about the images it is building, others should just be referenced in the helm charts only.
```
Chart.yaml
values.yaml
requirements.yaml # Declares local dependency charts, no repository field
templates/ # Chart templates
charts/
    subchart/
        Chart.yaml
        values.yaml
        templates/
    ...
```

Extra bonus: *all* application configuration is kept in Helm charts.

> Since there's no architecture diagram anywhere

https://skaffold.dev/docs/concepts/ ? 

> supposed to sit between dev and prod. It would seem it's mostly a wrapper that creates/applies/deletes all yaml files I have in a folder. It's not a full answer to a development environment.

In conjunction with `minikube`, `skaffold dev` builds & run all our containers in the configuration specified by our `helm` charts. If I run `skaffold deploy -p production` i.e. using the production profile, it will build & deploy a production version of our application?

> I have to search for how to set up ingresses with minikube and then search for how to add TLS to it. 
This should be in the docs IMO.

Why should `skaffold` behave differently based on the `Kubernetes` implementation? Why should your `Kubernetes` implementation be related?

We use `istio`'s `ingress-gateways` in production & staging, with `certmanager` for HTTPS with LetsEncrypt, and instead use `minikube`'s `ingress` in dev mode. Strictly speaking, the `ingress-gateways` actually interface with GCP's Load Balancer implementation. If we used AWS, that would be different.

> Then, the sync; 

We don't use so I cant comment.

I *love* the fact that skaffold does a simple clear set of tasks:
- Build specified targets using our build system.
- Deploy our configuration to our cluster, with the outputs of those builds.
- Offer the above in an automatic development loop.
- Customise the build & deployment based on a profile, i.e. dev vs prod.

**Comment by balopat** [comment]
@haf, thank you so much for the detailed feedback! 

> This is a review of the initial experience using you tool to manage this for dev:
> 
> * one API service
> * one backend batch service
> * one data database: eventstore
> * one metrics database: influxdb
> * prometheus

This is in scope - as long as you can deploy these things to Kubernetes, you should be able to spin it up with Skaffold. 

> The first experience is one of an MVP that ticks the most important box; a quick getting started doc that actually works. That said, the quick start seems to have displaced the architecture diagrams and overview docs; something DevSpace did really well.

Added https://github.com/GoogleContainerTools/skaffold/issues/1411 for architectural diagram. 

> After copying some staging yaml files from our k8s repo to the local dev repo, 

Can you elaborate on this a bit more, I'd be curious to understand your workflow more: what do you mean by k8s repo and local dev repo and "staging yaml files"? 

> I note that there's no guidance on how to think about helm charts for monitoring (which we need in dev), versus helm charts versus plain yaml files. Missing docs for this. Asking on [stackoverflow](https://stackoverflow.com/questions/53672777/deploy-external-helm-chart-with-skaffold/53677969#53677969) doesn't yield many responses, and the only one I've gotten so far doesn't help; I want to declare both ephemeral things (the containers with the code I change), and more stable things like Prometheus in a single file.

It's a great point that we could add recommendations in the docs around how to setup monitoring and other infrastructure. I'd be curious to see what worked for you as well. Added https://github.com/GoogleContainerTools/skaffold/issues/1412. 

> After getting the services up... (more or less; my deployment is properly rebuilt, but the backend statefulset doesn't; it tries to pull from my local GCP repo (which I have authenticated towards), despite skaffold actually building the image before it starts the statefulset — it would seem skaffold doesn't properly update the image at runtime for stateful set, but does properly do it for deployments)

This might be a bug - do you still see this issue? Do you have by any chance a simple reproduction? 

> ... I try to request the service; another thing DevSpace does much better — what use is a dev tool if I can't request things on it? I have to search for how to set up ingresses with minikube and then search for how to add TLS to it. This should be in the docs IMO.

Added an issue to cover guidance on service routing: https://github.com/GoogleContainerTools/skaffold/issues/1413

> Then, the sync; DevSpace spawns a shell inside a container — a nice touch — skaffold just starts it all and expect you to write down an exhaustive pattern match of files to sync. In my case, I want to keep the databases' state on host disk (like docker compose does with volumes); but there are no docs about this and nothing in the annotated skaffold yaml file either.

The filesync in Skaffold is literally just keeping up to date with small static files during development - it's not meant to be syncing database files. 
I'm not convinced yet that host folder mounting should belong to skaffold just yet. We can/should mention that in the infra docs. 

> * **The ability to run a script before building the dockerfiles**
Can you elaborate on what would you like to run before building the images? 

> * **The ability to run a script on e.g. 'api is healthy', 'db is healthy' to seed dev data/permissions**
This is another thing I would think we should tackle as documentation in the infra docs. First thought is that for simpler cases like init scripts, one time Jobs could execute that would be waiting for certain conditions should be sufficient.

/cc @dgageot

**Comment by haf** [comment]
Thanks for your reply,

> The filesync in Skaffold is literally just keeping up to date with small static files during development - it's not meant to be syncing database files.

Agree; that would be very hard to get correct. This problem is exacerbated on macOS because xhyve is deprecated (but had working mounts) and hyperkit doesn't give you working minikube mounts; but is recommended.

**Comment by haf** [comment]
> [StatefulSet] This might be a bug - do you still see this issue? Do you have by any chance a simple reproduction?

It may be related to that k8s requires a manual pod delete for it to replace the active image. I have a few stateful sets to provision today, so I'll look into it and report back more specifics.

**Comment by haf** [comment]
> Can you elaborate on this a bit more, I'd be curious to understand your workflow more: what do you mean by k8s repo and local dev repo and "staging yaml files"?

The staging environment is just a replica with a similar setup to production. I've got very similar yaml files for both of those, but for dev they are slightly different from the other two environments. I suppose a solution to this would be to state 'use helm with a variable file'.

> This might be a bug - do you still see this issue? Do you have by any chance a simple reproduction?

We've verified this works now, thanks.

It's still "pending" for mee to validate whether hyperkit successfully mounts directories.

**Comment by haf** [comment]
I think we can close this, since there's nothing more I have to add in here.
